File: Python/examples/AdvancedSpecificMultipleInstruments.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Advanced Specific Multiple Instruments
This example shows how to use specific channels ID and multiple connected
instruments

Requires at least two Fluigent pressure channels

Copyright (c) Fluigent 2019.  All Rights Reserved.
"""

# Print function for Python 2 compatibility
from __future__ import print_function 

from Fluigent.SDK import fgt_detect, fgt_init, fgt_close
from Fluigent.SDK import fgt_get_controllersInfo
from Fluigent.SDK import fgt_get_pressureChannelCount, fgt_get_pressureChannelsInfo
from Fluigent.SDK import fgt_get_pressure

# Detect all controllers
SNs, types = fgt_detect()
controllerCount = len(SNs)
print('Number of controllers detected: {}'.format(controllerCount))

# List all found controllers' serial number and type
for i, sn in enumerate(SNs):
    print('Detected instrument at index: {}, ControllerSN: {}, type: {}'\
          .format(i, sn, str(types[i])))

## Initialize specific instruments
# Initialize only specific instrument controllers here If you do not want
# a controller in the list or if you want a specific order (e.g. LineUP
# before MFCS instruments), rearrange parsed SN table
fgt_init(SNs)

# Get total number of initialized pressure channels
print('Total number of pressure channels: {}'.format(fgt_get_pressureChannelCount()))

## Get detailed information about all controllers

controllerInfoArray = fgt_get_controllersInfo()
for i, controllerInfo in enumerate(controllerInfoArray):
    print('Controller info at index: {}'.format(i))
    print(controllerInfo)

## Get detailed information about all pressure channels

pressureInfoArray = fgt_get_pressureChannelsInfo()
for i, pressureInfo in enumerate(pressureInfoArray):
    print('Pressure channel info at index: {}'.format(i))
    print(pressureInfo)

## Read pressure using unique ID
# If you want to address a specific channel, unique ID can be used. However
# if hardware changed channel may not be found

try:
    pressure1 = fgt_get_pressure(pressureInfoArray[0].indexID)
    print('Read pressure from ID {} : {:.2f}\n'.format(pressureInfoArray[0].indexID, pressure1))
except IndexError as e:
    print('WARNING: Cannot read pressure on channel 0')

try:
    pressure2 = fgt_get_pressure(pressureInfoArray[1].indexID)
    print('Read pressure from ID {} : {:.2f}\n'.format(pressureInfoArray[1].indexID, pressure2))
except IndexError as e:
    print('WARNING: Cannot read pressure on channel 1')

## Close the session
fgt_close()

File: C++/Examples/Advanced Specific Multiple Instruments/Advanced Specific Multiple Instruments.cpp
/*============================================================================
*               Fluigent Software Developement Kit for C++
*----------------------------------------------------------------------------
*         Copyright (c) Fluigent 2021.  All Rights Reserved.
*----------------------------------------------------------------------------
*
* Title:   Advanced Specific Multiple Instruments.cpp
* Purpose: This example shows how to use specific channels ID and multiple
*			connected instruments
*
* Software: fgt_SDK is the library managing all Fluigent instruments
*			"fgt_SDK_Cpp.h" is the wrapper to fgt_SDK dll
* Hardware setup: At least two Fluigent pressure modules
* Version: 21.0.0.0
* Date:	04/2021
*============================================================================*/


#include <iostream>
#include "fgt_SDK_Cpp.h"			// include wrapper to fgt_SDK dll, functions can also be accessed by loading the dll

int main()
{

	// Variables declaration
	fgt_CHANNEL_INFO pressure_info[256];		// Structure containing generic channel information
	fgt_CONTROLLER_INFO controller_info[256];
	fgt_INSTRUMENT_TYPE instrument_type[256];	// Enumeration of all types of supported instruments
	
	unsigned short SN[256];

	unsigned char controller_number = 0;		// Returned value of channel number
	unsigned char pressure_number = 0;
	float pressure;

	// Get total number , SN and type of detected Fluigent instrument controller
	controller_number = Fgt_detect(SN, instrument_type);
	std::cout << "Total number of controller(s): " << int(controller_number) << std::endl;

	// List all found controllers serial number and type
	for (unsigned char loop = 0; loop < controller_number; loop++)
	{
		std::cout << "Detected instrument at index: " << int(loop) << ", ControllerSN: " << SN[loop] << ", type: " << instrument_type[loop] << std::endl;
	}

	Fgt_initEx(SN);				// Initialize only specific instrument controllers here
								// If you do not want a controller in the list or if you want a specific order (e.g. LineUP before MFCS instruments)
								//	rearrange parsed SN table
	// Function Fgt_initEx can be called again, without the need of calling Fgt_close() before. 

	// Get total number of initialized pressure channel(s)
	Fgt_get_pressureChannelCount(&pressure_number);
	std::cout << "Total number of pressure channels: " << int(pressure_number) << std::endl;

	// Get detailed information about all controller(s)
	Fgt_get_controllersInfo(controller_info);
	for (unsigned char loop = 0; loop < controller_number; loop++)
	{
		std::cout << "Controller info at index: " << int(loop) << " SN: " << controller_info[loop].SN
			<< " Firmware: " << controller_info[loop].Firmware << " ID: " << controller_info[loop].id
			<< " type: " << controller_info[loop].InstrType << std::endl;
	}

	// Get detailed information about all pressure channels
	Fgt_get_pressureChannelsInfo(pressure_info);
	for (unsigned char loop = 0; loop < pressure_number; loop++)
	{
		std::cout << "Pressure channel info at index: " << int(loop) << " ,Controller SN: " << pressure_info[loop].ControllerSN
			<< " DeviceSN: " << pressure_info[loop].DeviceSN << " firmware: " << pressure_info[loop].firmware
			<< " position: " << pressure_info[loop].position << " index: " << pressure_info[loop].index
			<< " ID: " << pressure_info[loop].indexID << " InstrType: " << pressure_info[loop].InstrType << std::endl;
	}

	/** If you want to address a specific channel, unique ID can be used. However if hardware changed channel may not be found */

	// Read pressure
	Fgt_get_pressure(pressure_info[0].indexID, &pressure);
	std::cout << "Read pressure from ID " << pressure_info[0].indexID << ": " << pressure << std::endl;
	Fgt_get_pressure(pressure_info[1].indexID, &pressure);
	std::cout << "Read pressure from ID " << pressure_info[1].indexID << ": " << pressure << std::endl;

	Fgt_close();								// Close session

}

